woocommerce/woocommerce-gutenberg-products-block
# Allow coupons to be applied and removed from the cart

This is my first pass at implementing coupon functionality in the cart.

To allow cart data to keep in sync when things such as coupons are applied, this PR implements a new data store (`cart`) to store cart endpoint data. Two custom hooks exist to get data from the cart: `useStoreCart` and `useStoreCartCoupons`.

The main actions for coupon handling are `applyCoupon` and `removeCoupon`, but found in the new data store.

To handle errors, an error state is stored and rendered above the cart. This is a placeholder for now and will **need refactoring in a followup.**

Additionally, the "chip" links to remove coupons needs a new component and also needs addressing in a followup. I've added a placeholder button to ensure this is functional.

On the API side, to avoid the need to make a coupon request followed by a cart totals request, I've implemented 2 new endpoints to the cart API: `apply-coupon` and `remove-coupon`. Both return updated totals once used.

Closes #1747

### Screenshots

![2020-02-21 15 59 55](https://user-images.githubusercontent.com/90977/75050219-7cf4df00-54c3-11ea-90f9-5ce61fdc133e.gif)

### How to test the changes in this Pull Request:

1. Go to the cart
2. Apply a test coupon - the cart should update.
3. Remove the coupon - the cart should update.
4. Enter an invalid coupon - the cart should show a notice and the field should be cleared.
